### PR TITLE
add support for RGB colors from Discord, implements #8

### LINKED ISSUE #8: Support RGB colors from Discord
Right now we're mapping from RGB colors to the old hex codes. We can just pass the values over as straight RGB now instead of trying to translate.

### PR DESCRIPTION
**Pull request by michaeljecmen**
Again, this isn't really something that can be tested apart from manual end-to-end tests, so I'll drop the screenshots of the tests I ran locally to demonstrate this enhancement:

Test 1 (green):

Test 2 (pink):

Test 3 (gray):

Nothing too fancy here, just replaced the old hex formatting string with a java Color object.

### CODE REVIEW

File: src/net/peacefulcraft/cowbot/handlers/ModChatMessageHandler.java [hunk 1]
```diff
@@ -13,7 +13,7 @@ public static boolean isReplayMessage(String username, String message) {
 
   public static void handle(Message message) {
     if (!message.getAuthor().isPresent()) { return; }
-    String author = message.getAuthor().get().getUsername();;
+    String author = message.getAuthor().get().getUsername();
```
Comment (michaeljecmen) [inline src/net/peacefulcraft/cowbot/handlers/ModChatMessageHandler.java:16]: also removed the double semicolon here because I noticed it and couldn't help myself

**Comment by michaeljecmen** [comment]
Alright, the parser is about as good as it's going to get without a lot more complexity. I added a testing module and a bunch of unit tests for the parsing code. I think my comments are already pretty good if you want to understand what it's doing, but feel free to ask any questions you may have.

The parser handles Discord formatting for italics (text wrapped with \* or \_), bold (text wrapped with \*\*), underlines (text wrapped with \__), strikethrough (text wrapped with \~~), and obfuscation (text wrapped with \||). You can also escape those characters with backslash \\ if you'd like to actually use them. In general, apart from certain edge cases which involve multiple of the same formatting code (how should \*\*\*A\*\*B\* be parsed? "\***A***B*" or "***A**B*"?), this parser behaves exactly like Discord's, and only takes a one-pass approach.

Furthermore, (and if any users are reading this), if you're considering using \* based italics alongside \*\* based boldface, you can simply replace the italics with the \_ version to ensure you'll get the same formatting you see in Discord. The same applies for \_ based italics and \_\_ based underlines -- just use \* based italics to not confuse the parser. These are the only cases for which I could cause different behavior, and in the worst case you'll just see some formatting codes in your Minecraft output.

Feel free to build on this to attempt to correct for these differences, but I'm unable to at the moment. All in all I think this is a super neat feature, and I'm glad it was a part of the PR.

File: src/test/java/TestDiscordToMinecraftFormattingTranslator.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        assertTrue(italicsUnd.empty());
+
+        assertEquals(7, italicsAst.peek().start);
+        assertEquals(9, italicsAst.peek().end);
+        italicsAst.pop();
+        assertTrue(italicsAst.empty());
+    }
+
+    // @Test
+    // public void testOverlappingItalicsBold0() {
```
Comment (michaeljecmen) [inline src/test/java/TestDiscordToMinecraftFormattingTranslator.java:180]: This is an example of a difference between the DiscordToMinecraftFormattingTranslator's parsing and the official Discord parsing. This test case fails, but I've left it in (commented out) in case a future contributor wants to modify my logic to somehow handle this case. I think I'd have to fundamentally change the parsing approach in order to account for this, but as discussed in the conversation tab I don't think it's that important to have 1:1 perfect Discord formatting representation.

**Comment by michaeljecmen** [comment]
Ok, one final note: I added some more tests until I achieved 100% line and branch coverage for parse() and translate(). Here's the report:
image

If you would like, I can push the changes to the pom which allow coverage report generation, but this was more for me to test the adequacy of my tests -- not something I see as particularly beneficial to other modules in CowBot.

File: src/net/peacefulcraft/cowbot/handlers/GameChatMessageHandler.java [hunk 3]
```diff
@@ -8,32 +8,39 @@
...
 public class GameChatMessageHandler {
 
   public static void handle(Message message) {
     String author = message.getAuthor().get().getUsername();
     Member sender = message.getAuthor().get().asMember(message.getGuild().block().getId()).block();
     Color color = sender.getColor().block();
-    String roleHex = String.format("#%02X%02X%02X", color.getRed(), color.getBlue(), color.getGreen());
+
+    // fully qualified name to avoid name collisions with discord4j Color
+    java.awt.Color rankColor = new java.awt.Color(color.getRed(), color.getGreen(), color.getBlue());
```
Comment (michaeljecmen) [inline src/net/peacefulcraft/cowbot/handlers/GameChatMessageHandler.java:21]: The only changes pertaining to the colored names feature are here, the new line 30, and the new line 43. The rest of the changes pertain to the text formatting feature.

**Review by Parsonswy — CHANGES_REQUESTED** [review]
Can you add tooling to `.vscode/` so that VSCode recognizes there are tests and can run them? This repo is a bit behind in our tooling, but the [template](https://github.com/peacefulcraft-network/Bukkit-Plugin-Template/tree/main/.vscode) has extension suggestion. I am not sure what all is needed after that to get the test runner to recognize the test cases.

Otherwise this looks good and we appreciate the time and thoughtful approach you've put into this 🙂.

**Comment by michaeljecmen** [comment]
Hey @Parsonswy, I've spent quite a lot of time on this to no avail -- I can get the extension working to where you can run the tests from VSCode, but it requires that the junit dependency no longer be test scoped, which is obviously undesirable. Furthermore, although maven can build the project via the VSCode toolbar in the bottom left (and through ```mvn compile``` in the terminal), the test runner complains about build errors each time you run the tests. The tests still work, and pass, but you have to click through the build error every time. Also, you can't debug :), it just doesn't hit any breakpoints (probably due to the build errors).

Long story short, I'm not sure you want me to push this code -- based on the number of internet searches I've made, a lot of people seem to have trouble with this sort of thing and there are a lot of unresolved forum posts about this issue exactly. I've pushed what I got to https://github.com/michaeljecmen/CowBot/tree/tooling if you want to pull that and see if it works on your end. Barring that I'm pretty stuck at the moment.

**Comment by Parsonswy** [comment]
> 

Alright. If it is a non-trivial amount of work it is not worth the time. I can just setup CI for it here, or take 4 clicks to open a terminal window.

**Review by Parsonswy — APPROVED** [review]
